pkgs/top-level/all-packages.nix

```diff
@@ -29471,6 +29471,12 @@ in
 
   rocksndiamonds = callPackage ../games/rocksndiamonds { };
 
+  rott = callPackage ../games/rott { };
+
+  rott-shareware = callPackage ../games/rott {
+    buildShareware = true;
+  };
```

Comment:
```suggestion
  rott-shareware = rott.override {
    buildShareware = true;
  };
```

Reply:
is this really better/clearer? I personally see overrides as "necessary evil". I  also use them sometimes, and in some cases it's unavoidable because there's no better way, but in the top-level expression I think it's a different story.

I believe that the convention that we follow in Nixpkgs is that every package declares a function, in which the function header specifies the dependencies and other parameters that allow you to compose multiple variants of a package.

In the top-level expression we basically invoke these functions with the right parameters. In more than 95% of the cases, there's only one single variant. As a result, a simple invocation to `callPackage {}` suffices that auto composes it.

However, if we want to compose different variants, I believe the practice should be to simply call the function with different parameters, rather than taking another composition and then overriding it.

Furthermore, if I'm not mistaken, according to @edolstra overrides are also bad because they cause memory leaks. So that's another reason to avoid them if we can, and this particular situation we can.

So IMO my variant that just composes a package with difference parameters should be better.

I'm not entirely sure if we have agreed to always by the community compose variants by overrides. If this breaks the established conventions, then I'll happly change it, but personally I believe we shouldn't use an override here.

What do you think?

Reply:
I've manually scanned the `all-packages.nix` file and it seems that we're using both strategies, but the `override` strategy seems to be far more common.  For example, `logstash` is still a package that uses a function invocation to compose variants rather than overrides.

Not sure, how this strategy came to be and if there's another reason that I may be missed. Maybe overlays require them?

I'm of course willing to change this to an override, since that's more common, although I still think a different function invocation is better.

Not sure if @edolstra has an opinion about this with regards to memory usage.

Reply:
*Having* a `.override` attribute causes memory leaks (since it prevents the inputs to a function from being garbage-collected). But given that we have it, we may as well use it...

Reply:
Using override is preferred because if you change the path of the package definition, callPackage or add some new inputs you don't need to update two places. For example if we would override an input with an older library to fix a failing build the second package could easily be forgotten.

Reply:
Well both approaches seems to have their pros and cons -- with an override you don't have to refer to the same file multiple times, but we introduce a memory leak, and with a function invocation we prevent a memory leak but we need to update two files.

I'll change this an override, because I guess since this is used more often, this has become the established practice. I still prefer the other solution, but I guess we can always change this into something better when the memory leaks have been sorted out.